### ISSUE TITLE
Mutation fails if Object subcolumn is used in WHERE

### ISSUE THREAD

**Issue by davenger** [ClickHouse/ClickHouse#49738]
Repro steps:

```
DROP TABLE IF EXISTS t_obj SYNC;

SET allow_experimental_object_type=1;

CREATE TABLE t_obj(id Int32, name Object('json')) ENGINE = MergeTree() ORDER BY id;

INSERT INTO t_obj select number, '{"a" : "' || toString(number) || '"}' FROM numbers(100);

SELECT COUNT() FROM t_obj WHERE name.a == '42';

ALTER TABLE t_obj DELETE WHERE name.a == '42' SETTINGS mutations_sync=2;

SELECT COUNT() FROM t_obj;


DROP TABLE t_obj SYNC;
```

Error:
```
[davenger-xps] 2023.05.10 16:53:07.379915 [ 2469798 ] {14d38d5c-7fec-42c6-bbeb-c7b92ce984d3} <Error> executeQuery: Code: 341. DB::Exception: Exception happened during execution of mutation 'mutation_3.txt' with part 'all_1_1_0_2' reason: 'Code: 47. DB::Exception: Missing columns: 'name.a' while processing query: 'SELECT count() WHERE name.a = '42'', required columns: 'name.a', maybe you meant: 'name'. (UNKNOWN_IDENTIFIER) (version 23.5.1.1)'. This error maybe retryable or not. In case of unretryable error, mutation can be killed with KILL MUTATION query. (UNFINISHED) (version 23.5.1.1) (from 127.0.0.1:33222) (comment: 02352_lightweight_delete_and_object_column.sql) (in query: ALTER TABLE t_obj DELETE WHERE name.a == '42' SETTINGS mutations_sync=2;), Stack trace (when copying this message, always include the lines below):

0. /home/davenger/src/ClickHouse/contrib/llvm-project/libcxx/include/exception:134: std::exception::capture() @ 0x000000002be6f722 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
1. /home/davenger/src/ClickHouse/contrib/llvm-project/libcxx/include/exception:112: std::exception::exception[abi:v15000]() @ 0x000000002be6f6db in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
2. /home/davenger/src/ClickHouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x0000000067d3b000 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
3. /home/davenger/src/ClickHouse/src/Common/Exception.cpp:90: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000004068851e in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
4. /home/davenger/src/ClickHouse/src/Common/Exception.h:54: DB::Exception::Exception(String&&, int, bool) @ 0x000000002be526e7 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
5. /home/davenger/src/ClickHouse/src/Common/Exception.h:81: DB::Exception::Exception<char const*, String, String&, String&>(int, FormatStringHelperImpl<std::type_identity<char const*>::type, std::type_identity<String>::type, std::type_identity<String&>::type, std::type_identity<String&>::type>, char const*&&, String&&, String&, String&) @ 0x0000000059c21fb3 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
6. /home/davenger/src/ClickHouse/src/Storages/MergeTree/MergeTreeMutationStatus.cpp:26: DB::checkMutationStatus(std::optional<DB::MergeTreeMutationStatus>&, std::set<String, std::less<String>, std::allocator<String>> const&) @ 0x0000000059c2178c in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
7. /home/davenger/src/ClickHouse/src/Storages/StorageMergeTree.cpp:569: DB::StorageMergeTree::waitForMutation(long, String const&, bool) @ 0x000000005a2c29d8 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
8. /home/davenger/src/ClickHouse/src/Storages/StorageMergeTree.cpp:539: DB::StorageMergeTree::waitForMutation(long, bool) @ 0x000000005a2bd501 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
9. /home/davenger/src/ClickHouse/src/Storages/StorageMergeTree.cpp:595: DB::StorageMergeTree::mutate(DB::MutationCommands const&, std::shared_ptr<DB::Context const>) @ 0x000000005a2c61e5 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
10. /home/davenger/src/ClickHouse/src/Interpreters/InterpreterAlterQuery.cpp:160: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x0000000055cbcf63 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
11. /home/davenger/src/ClickHouse/src/Interpreters/InterpreterAlterQuery.cpp:64: DB::InterpreterAlterQuery::execute() @ 0x0000000055cb92ef in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
12. /home/davenger/src/ClickHouse/src/Interpreters/executeQuery.cpp:717: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000057232a1d in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
13. /home/davenger/src/ClickHouse/src/Interpreters/executeQuery.cpp:1173: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x000000005722ab62 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
14. /home/davenger/src/ClickHouse/src/Server/TCPHandler.cpp:420: DB::TCPHandler::runImpl() @ 0x000000005a86bd42 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
15. /home/davenger/src/ClickHouse/src/Server/TCPHandler.cpp:2043: DB::TCPHandler::run() @ 0x000000005a8924ea in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
16. /home/davenger/src/ClickHouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x000000006789fae5 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
17. /home/davenger/src/ClickHouse/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x00000000678a068d in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
18. /home/davenger/src/ClickHouse/base/poco/Foundation/src/ThreadPool.cpp:188: Poco::PooledThread::run() @ 0x0000000067e706d4 in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
19. /home/davenger/src/ClickHouse/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x0000000067e690ea in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
20. /home/davenger/src/ClickHouse/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000067e6690f in /home/davenger/src/ClickHouse/build_Debug/programs/clickhouse
21. ? @ 0x00007ffff7c94b43 in ?
22. ? @ 0x00007ffff7d26a00 in ?

```

**Comment by alexvaut** [comment]
Same on my side... Any fix or workaround ahead ?

**Comment by Algunenano** [comment]
The reproducer works fine in 23.10 and master and it's an experimental feature that's getting deprecated.

**Comment by alexvaut** [comment]
Hi @Algunenano,
Thanks for the prompt answer. However I'm using 23.10.3 and I'm getting the same error. What do you mean by "it's an experimental feature that's getting deprecated" ? Deleting based on json content is not experimental, is it ?

**Comment by Algunenano** [comment]
```
$ cat repro.txt 
DROP TABLE IF EXISTS t_obj SYNC;

SET allow_experimental_object_type=1;

CREATE TABLE t_obj(id Int32, name Object('json')) ENGINE = MergeTree() ORDER BY id;

INSERT INTO t_obj select number, '{"a" : "' || toString(number) || '"}' FROM numbers(100);

SELECT COUNT() FROM t_obj WHERE name.a == '42';

ALTER TABLE t_obj DELETE WHERE name.a == '42' SETTINGS mutations_sync=2;

SELECT COUNT() FROM t_obj;


DROP TABLE t_obj SYNC;
```

* 23.9:
```
/mnt/ch/official_binaries/clickhouse-common-static-23.9.1.1854/usr/bin/clickhouse local --queries-file repro.txt 
1
Code: 341. DB::Exception: Exception happened during execution of mutation 'mutation_2.txt' with part 'all_1_1_0' reason: 'Code: 47. DB::Exception: Missing columns: 'name.a' while processing query: 'SELECT count() WHERE name.a = '42'', required columns: 'name.a', maybe you meant: 'name'. (UNKNOWN_IDENTIFIER) (version 23.9.1.1854 (official build))'. This error maybe retryable or not. In case of unretryable error, mutation can be killed with KILL MUTATION query. (UNFINISHED)
```

23.10:
```
$ /mnt/ch/official_binaries/clickhouse-common-static-23.10.1.1976/usr/bin/clickhouse local --queries-file repro.txt 
1
99
```

That reproducer works.


> Deleting based on json content is not experimental, is it ?

The `Object('json')` feature is experimental. Any usage related to it is then, by definition, experimental.

**Comment by alexvaut** [comment]
My bad, once more "latest" version of my docker container proved nme wrong since I was using an old clickhouse docker image already on my machine. I confirm it's fixed in 23.10.3
